### PR TITLE
[url_launcher_windows] Support win32 and UWP hosts

### PR DESCRIPTION
**Pull request by clarkezone**
This PR adds support for the winuwp target.

https://github.com/flutter/flutter/issues/70207
https://github.com/flutter/flutter/issues/14967

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format](../script/tool/README.md#format-code))
- [X] I signed the [CLA].
- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by clarkezone** [comment]
FWIW I noticed last night that licenses needed updating but didn't have time to locate the licenses file

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
High level before reviewing the code itself:
1. This appears to break Win32 builds. I can't see why (https://github.com/flutter/flutter/issues/81297) but hopefully it's reproducible locally.
2. This needs to actually be built in CI.
3. Relatedly, this needs to be running the integration tests.

For `2`, what's the current status of UWP runners? Does `flutter create .` work? Is the template still unstable? If Yes and Yes, we can add the code to create it on the fly that we used to use for other desktop builds here.

For `3`, is there a LUCI recipe that can run UWP apps yet?

**Comment by stuartmorgan** [comment]
> FWIW I noticed last night that licenses needed updating but didn't have time to locate the licenses file

The correct license form for this repo is what was in `url_launcher_plugin.cpp` before you changed it. You just need to revert that change, and use that text in the other files.

**Comment by clarkezone** [comment]
1. The win32 build of url_launcher example works fine locally so not sure how to reason over the failure.  Not sure if the infra is also trying to run in UWP mode? 
2. Flutter create isn't ready yet but in progress so this one is currently blocked
3. I don't believe so.  This would likely need WinAppDriver support (https://github.com/Microsoft/WinAppDriver) in LUCI.  @cbracken for thoughts on this

**Comment by stuartmorgan** [comment]
> 1. The win32 build of url_launcher example works fine locally so not sure how to reason over the failure.  Not sure if the infra is also trying to run in UWP mode?

I'm not clear on what this means. How would `flutter build windows` be UWP mode?

**Comment by clarkezone** [comment]
> > 1. The win32 build of url_launcher example works fine locally so not sure how to reason over the failure.  Not sure if the infra is also trying to run in UWP mode?
> 
> I'm not clear on what this means. How would `flutter build windows` be UWP mode?

flutter build windows is not uwp.  if that is what CI is doing (only), then it works file locally and hence i cannot reason over why CI is failing.

File: packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h [hunk 1]
```diff
@@ -4,7 +4,14 @@
 #ifndef PACKAGES_URL_LAUNCHER_URL_LAUNCHER_WINDOWS_WINDOWS_INCLUDE_URL_LAUNCHER_WINDOWS_URL_LAUNCHER_PLUGIN_H_
 #define PACKAGES_URL_LAUNCHER_URL_LAUNCHER_WINDOWS_WINDOWS_INCLUDE_URL_LAUNCHER_WINDOWS_URL_LAUNCHER_PLUGIN_H_
 
+#include <flutter/method_channel.h>
+#include <flutter/plugin_registrar_windows.h>
+#include <flutter/standard_method_codec.h>
 #include <flutter_plugin_registrar.h>
+#include <winrt/Windows.Foundation.h>
```
Comment (stuartmorgan) [inline packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h:11]: Perhaps this is what's breaking Win32 compilation, if this header isn't available on the bots? I'm not clear on why this has been added in the first place; perhaps remove it and see what happens.

Comment (clarkezone) [reply packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h]: Good point.. pushed a change to test this theory.

File: packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h [hunk 2]
```diff
@@ -23,4 +30,12 @@ FLUTTER_PLUGIN_EXPORT void UrlLauncherPluginRegisterWithRegistrar(
 }  // extern "C"
 #endif
 
+void OpenLink(
+    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> value,
+    std::string url_string);
```
[the remark this answers is not in the document]
Comment (clarkezone) [reply packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h:34]: Sheesh you are right as ever.  Will be fixed in next commit.

**Comment by clarkezone** [comment]
Seeing this in the latest run..

`Checking Dart SDK version... 
Downloading Dart SDK from Flutter engine ... 
Building flutter tool... 
Running pub upgrade... 
Got TLS error trying to find package args at https://pub.dartlang.org.
Error (69): Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left) 
Running pub upgrade... 
Got TLS error trying to find package html at https://pub.dartlang.org.
Error (69): Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (8 tries left) 
Running pub upgrade... 
Got TLS error trying to find package flutter_template_images at https://pub.dartlang.org.
Error (69): Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (7 tries left) 
Running pub upgrade... 
Terminate batch job (Y/N)? 
^CError (-1073741510): Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (6 tries left) 
Running pub upgrade... `

**Comment by cbracken** [comment]
@clarkezone this is likely the result of https://github.com/flutter/flutter/issues/81837

**Comment by stuartmorgan** [comment]
@clarkezone Could you merge in master and see what happens? I've temporarily replaced the currently-completely-broken Windows LUCI CI with GA-based CI and I'm curious if it will give us different results.

**Comment by clarkezone** [comment]
Will do

**Comment by clarkezone** [comment]
Merged in, results below

**Comment by stuartmorgan** [comment]
That passed, which is good; that means it's not something like reliance on local state on your machine to compile successfully. Now that the new LUCI bots are online, you can sync in master yet again and retry, and we can see if the new bots are free of whatever the mystery issue was 🤞🏻

**Comment by stuartmorgan** [comment]
Success! Looks like the failure was something specific to the old bots, and the new bots don't have the same issue. One blocker down :)

@cbracken can you ping me here when `flutter create` works for UWP?

File: packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (clarkezone) [reply packages/url_launcher/url_launcher_windows/windows/include/url_launcher_windows/url_launcher_plugin.h]: Fixed.

**Comment by clarkezone** [comment]
CR feedback addressed

**Comment by domesticmouse** [comment]
@stuartmorgan I believe `flutter create` now works on Flutter master channel, once `flutter config --enable-windows-uwp-desktop` has been run.

**Comment by stuartmorgan** [comment]
I tried building locally, and hit compile warnings (treated as errors) in cppwinrt/winrt/base.h (SDK version 18362). I'll try upgrading my SDK version, but if we're not going to require a higher SDK version for UWP (`doctor` didn't complain) we should adjust the build flags in the CMake file here.

**Comment by clarkezone** [comment]
Do u have UWP workload installed along with c++ support for UWP?

**Comment by stuartmorgan** [comment]
I'll have to check. Is that's the issue, then doctor will definitely need an update.

**Comment by clarkezone** [comment]
Yea it's on the list @cbracken @jonahwilliams

**Comment by stuartmorgan** [comment]
Tomorrow I'll file an issue laying out the pieces still left to unblock plugins landing here, for tracking.

**Comment by stuartmorgan** [comment]
https://github.com/flutter/flutter/issues/82822 tracks the blockers. We can add more there if we think of anything else.

**Comment by stuartmorgan** [comment]
There's still no `flutter drive` support, but this can be rebased against master where we have build support for UWP plugins to ensure that everything is building as expected.

**Comment by clarkezone** [comment]
I just got unblocked on building engine, i will rebase on push this sometime as soon as i can

**Comment by stuartmorgan** [comment]
I realized that my restructure of the plugin to make it testable caused a bunch of non-trivial conflicts, so I went ahead and resolved it. The resolution is hacky (ifdefs in a header, splitting class implementation between several files) so it'll need cleanup, but it's at least a cleaner starting point that the pile of conflicts.

I haven't tested the UWP part since I still don't have that working on my Windows machine, so apologies if I broke that. But we can at least see if CI builds it :)

**Comment by stuartmorgan** [comment]
Hm, the UWP build timed out on CI, which is what happened to me locally. I thought that was just something about my setup, but apparently not.

**Comment by clarkezone** [comment]
Interesting.

**Comment by clarkezone** [comment]
I'm blocked currently due to inability to build the UWP target on my machine.

**Comment by stuartmorgan** [comment]
@cbracken @clarkezone Do we have a path to getting this unblocked, or should we close this out for now?

**Comment by clarkezone** [comment]
I got unblocked before the holidays but haven't had a chance to look at this since then, appols.

**Comment by cbracken** [comment]
I have this and a couple other plugins PRs at the top of my review queue now that I’m back from holidays.

**Comment by stuartmorgan** [comment]
> Looks like there’s some work to do first though?

Yes, this PR currently hangs UWP builds (at least for me and CI). It's possible I broke it with my update to the PR, but IIRC I already had that problem.

**Comment by cbracken** [comment]
@clarkezone is this something that you expect you'll pick up any time soon? If not, shall we close it?

**Comment by clarkezone** [comment]
I aspire to landing this but I haven't done a good job at making any progress.  If you'd be happier closing it out I'm OK with that.

**Comment by cbracken** [comment]
SGTM. Closing for now. When you get time and it's ready for review; fire a new PR our way!

**Pull request closed without merging**